### DISCUSSION TITLE
Mocking svelte component

### DISCUSSION THREAD

**Discussion by zsaquarian** [vitest-dev/vitest#869] [Q&A] [answered]
Hi!
Does anyone know how to mock/stub a svelte component? I want to separate my unit tests from each other.

In jest you would use something like [svelte-mock](https://github.com/kobejean/svelte-mock) as your transformer, but how would you do it in vitest?

Thanks in advance!

**Comment by sheremet-va** [comment, marked as answer]
What if you just don't add transformers, and use everything else? Have you tried it?

**Reply by zsaquarian** [reply]
When I try to do
```typescript
import { vi, it } from 'vitest';
import { render } from '@testing-library/svelte';
import ComponentToTest from '@components/ComponentToTest.svelte';

vi.mock('@components/ComponentToMock.svelte');

it('should mock', () => {
  render(ComponentToTest);
}
```

I get the error: `TypeError: Cannot read property 'fragment' of undefined`

And if i replace `vi.mock('@components/ComponentToMock.svelte');` with `vi.mock('@components/ComponentToMock.svelte', () => SomeMockComponent);`, then I get: `TypeError: __vite_ssr_import_1__.default is not a constructor
`

**Reply by sheremet-va** [reply]
> When I try to do
> 
> ```ts
> import { vi, it } from 'vitest';
> import { render } from '@testing-library/svelte';
> import ComponentToTest from '@components/ComponentToTest.svelte';
> 
> vi.mock('@components/ComponentToMock.svelte');
> 
> it('should mock', () => {
>   render(ComponentToTest);
> }
> ```
> 
> I get the error: `TypeError: Cannot read property 'fragment' of undefined`
> 
> And if i replace `vi.mock('@components/ComponentToMock.svelte');` with `vi.mock('@components/ComponentToMock.svelte', () => SomeMockComponent);`, then I get: `TypeError: __vite_ssr_import_1__.default is not a constructor `


You would need to do `vi.mock('@components/ComponentToMock.svelte', () => ({ default: SomeMockComponent });` if you want to mock default

**Reply by zsaquarian** [reply]
> > When I try to do
> > ```ts
> > import { vi, it } from 'vitest';
> > import { render } from '@testing-library/svelte';
> > import ComponentToTest from '@components/ComponentToTest.svelte';
> > 
> > vi.mock('@components/ComponentToMock.svelte');
> > 
> > it('should mock', () => {
> >   render(ComponentToTest);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get the error: `TypeError: Cannot read property 'fragment' of undefined`
> > And if i replace `vi.mock('@components/ComponentToMock.svelte');` with `vi.mock('@components/ComponentToMock.svelte', () => SomeMockComponent);`, then I get: `TypeError: __vite_ssr_import_1__.default is not a constructor `
> 
> You would need to do `vi.mock('@components/ComponentToMock.svelte', () => ({ default: SomeMockComponent });` if you want to mock default

It works! Thanks!

**Reply by MahmudovaA** [reply]
@sheremet-va is it possible to do something like this
```
vi.mock('@components/ComponentToMock.svelte', () => ({
  default: () => <div>mocked</div>
}));
```
(as mentioned [here](https://github.com/vitest-dev/vitest/discussions/2900#discussioncomment-5078929)) for a Svelte component? 
I have a lot of nested components that I want to mock, but I prefer not to create `SomeMockComponent` for each one of them. Thanks in advance.
